### DISCUSSION TITLE
City Hub Plug and Play Options

### DISCUSSION THREAD

**Discussion by Ben-Toolbox** [pybricks/support#1659] [Q&A] [answered]
Hi all,

I have been working on some code to give the City Hub some more plug and play flexibility in the younger LEGO Engineering classes I teach.  

Before Pybricks, in my classes I simply used the LEGO firmware and connected the hubs to the LEGO remotes.  I was hoping to have at least 2 non-remote plug and play options that would be automatically determined by what devices are plugged into the hub:

- Option 1 - Hub as a battery pack with direction change (if only motors are plugged in)
I wrote a code that allowed the hub to work as a battery pack, with a tap of the button turning the motor on or off, and a button hold (more than a second) triggering a direction change of the motor.  
It seems to currently work fine (though I am sure it could use some work)

![pybricks_program_2024-06-03T00_23_49](https://github.com/pybricks/support/assets/171524495/1e1e355d-9f2e-4e9d-81d3-0b377039d32f)

- Option 2 - Tilt sensor controlled motor (if tilt sensor and motor are plugged in)
This code is much simpler and would have the motor continually run at a power determined by the tilt sensor pitch (with a bit of a multiplier)

![pybricks_program_2024-06-03T02_05_37](https://github.com/pybricks/support/assets/171524495/645380d5-891d-4e70-829b-4555d159c0ce)

As I mentioned above, I am hoping there could be a way I could figure out a single set code that could run either of these programs depending on what devices are plugged in?   The kids in the class are quite young, so I am avoiding any kind of program selection using the single hub button.  

Thanks!

**Comment by laurensvalk** [comment, marked as answer]
Thanks Ben for using the forum!

To select either of these block programs automatically based on what is plugged in, you can do something [like this technique](https://pybricks.com/project/spike-hub-menu/#choose-which-program-to-run-based-on-what-is-plugged-in):

```python
try:
    # This will run if the tilt sensor and the simple motor are plugged in.
    import your_sensor_block_program
except OSError:
    # Otherwise it will run this.
    import your_button_block_program
```

However, I think you'd like to take this further and have it automatically restart when you change the cables as you go, right?

Or would it be OK to restart the script in that case, by pressing the hub button again?

**Comment by Ben-Toolbox** [comment]
Thanks Laurens!

This looks perfect - I'll give that a try.  Sorry - I must have missed this last section of that page!

And yes, ideally it would be wonderful to have some kind of an ongoing reinitialization of the ports so that the program can change on the fly depending on what is plugged in.

As an additional condition, I was wondering about remote connection.  When the hub starts up, could you allow a short amount of time to connect to the remote... then:
-if it connects, a simple remote control program boots up.
-if no remote connects, then the programs I was referring to above kick in depending on what is plugged in.

Thanks so much!
